バックアップスクリプトの作成

logica0419 opened the pull request:
とりあえずローカルでファイルの転送(一応バックアップ)ができるようになりました
環境変数は今現在.envから読んでるので、後々変えるかも
とりあえず次はGCPへの接続をしたいと思います

Commit:
.gitignore、go moduleを導入

Commit:
backup.goをmain.goに変更

Commit:
環境変数の読み込みをできるようにした

Commit:
ローカルにおいてファイルを別フォルダに転送(バックアップ)できるようになった

Commit:
変数名・ログをわかりやすいように変更

Commit:
GCP関係の処理をGCP.goに移動

Commit:
・GCPライブラリを導入
・サービスアカウントと紐づけたクライアントが作成できるように

Commit:
・コードに説明を追加
・バケットの作成を実装

Commit:
snappyを使わない、非圧縮のバックアップを実装

logica0419 commented:
ローカルのフラットなディレクトリの中にある全ファイルを、GCPのプロジェクト(jsonの鍵で指定)上、Cloud Strageのバケットに非圧縮でバックアップできるようになりました

Commit:
snappyをつかった圧縮を実装

logica0419 commented:
snappyまで使ってローカルのファイルをバックアップできました。
環境変数の取得方法・エラーが出たとき実行を中断するか・Webhookがこれから詰めていかなきゃいけない部分ですが、とりあえずレビューをお願いしたいと思います。

hijiki51 reviewed (commented):
goのfile名はsnake_caseで扱うことが多いのでGCP.goもそうしたほうがいいです。

(goはxxx_test.goをテスト用のコードとして認識するのでその関係だと思います)

hijiki51 commented on GCP.go:
```diff
@@ -0,0 +1,95 @@
+	"google.golang.org/api/option"
+)
+
+func create_client(GCPKey string) *storage.Client {
```
goではpublicなものはPascalCase,privateなものはcamelCaseで記述するようになっているのでそのようにしてください

https://golang.org/ref/spec#Exported_identifiers

logica0419 replied:
リネームします

hijiki51 commented on GCP.go:
```diff
@@ -0,0 +1,95 @@
+		log.Fatal(err)
+	}
+
+	return bucket, fmt.Sprintf("Bucket \"%s\" successfully created", bucketName)
```
成功時のログのメッセージは戻す必要はないので戻さずにここで処理していいとおもいます

logica0419 replied:
なるほど、成功時はnilを返せばいいということで合ってますかね？

hijiki51 replied:
nilを返すというのが、errをかえすようにして、成功時にはnilを返すようにするという意味ならそうです

hijiki51 commented on GCP.go:
```diff
@@ -0,0 +1,95 @@
+	// jsonで渡された鍵のサービスアカウントに紐づけられたクライアントを建てる
+	client, err := storage.NewClient(ctx, option.WithCredentialsFile(GCPKey))
+	if err != nil {
+		log.Fatal(err)
```
goではerrを返して呼び出し元でその処理をどう扱うかを決めることが一般的なので、ここでFatalして中断するのではなく、errをそのまま返してあげてください

logica0419 replied:
わかりました

hijiki51 commented on GCP.go:
```diff
@@ -0,0 +1,95 @@
+package main
```
パッケージは分割したほうがいいと思います

logica0419 replied:
分割しておきます

hijiki51 reviewed (commented):
レビューの作成に失敗したので再投稿です、
ごめんなさい

Commit:
レビューの適応とwebhookパッケージの追加
・gcp.goのパッケージ化
・関数 / ファイル名のリネーム
・関数がエラーを返し、成功時には関数内部でログを流すように
・webhookパッケージの作成
・webhook用のメッセージ(仮)を生成するように

Commit:
レビューの適応とwebhookパッケージの追加
・gcp.goのパッケージ化
・関数 / ファイル名のリネーム
・関数がエラーを返し、成功時には関数内部でログを流すように
・webhookパッケージの作成
・webhook用のメッセージ(仮)を生成するように

logica0419 commented:
レビューの内容を反映させ、改善しました
再レビューで問題なさそうか確認お願いします

Commit:
修正のためのマージコミットを作成
ローカルでプッシュ済みのコミットに重ねがけしてしまい、リモートとコンフリクト起こしたので…

Commit:
環境変数の読み込みを改善
関数では読み込み方を指定するだけとした

Commit:
環境変数の読み込みをパッケージごとのグローバル変数への代入に変更

Commit:
traQ Webhook BotへHTTP POSTリクエストを送る部分の実装
一応これにて全機能実装完了

Commit:
Merge branch 'master' of https://github.com/traPtitech/server-backup into backup

Commit:
README.mdを編集
全体的に体裁を整えた

logica0419 commented:
必要な機能の実装はすべて終えました。
・Webhookのメッセージは、今のところConohaオブジェクトストレージのバックアップスクリプトをほぼ丸コピしていますが、これ以外に表示した方がいい情報、表示しなくていい情報はありますか？
https://q.trap.jp/messages/7112823f-3688-4287-b08e-4b66190fb535
・環境変数の読み込みはどのような形で行うのでしょうか？いまのところ同じディレクトリにRubyの.envを置いてそれを読み込んでいますが…
という2点が相談しなきゃいけないところかな。
レビューの方、よろしくお願いします。

Commit:
コードの微修正(順序など)

FujishigeTemma commented:
ありがとうございます！表示は今のもので大丈夫だと思います！
環境変数は https://git.trap.jp/SysAd/tokyotech.org/src/branch/master/roles/conoha-backup-helper/templates/conoha-backup-helper.service.j2#L8 みたいに指定するので今のままで問題ないです:ok_hand:

logica0419 commented:
わかりました！ありがとうございます。
ただ、今のままだと.envが見つからないときエラー吐いて強制終了するので、.envなくてもそのまま実行できるようにしておきます

sapphi-red commented:
ローカルでテストするときも`.env`使わなくても環境変数セットして実行するようなsh書けばいいと思うので、`.env`周りの記述は消しちゃったほうがよさそうです

sapphi-red commented on README.md:
```diff
@@ -1,17 +1,22 @@
-# server-backup
+# traQ_local-backup-helper
```
traQに限らず特定のフォルダ以下のバックアップに利用できると思うのでそういう感じに説明を変えてほしいです :pray:

sapphi-red commented on go.mod:
```diff
@@ -0,0 +1,15 @@
+module modset
```
`module github.com/traPtitech/リポジトリ名`に変更してほしいです

sapphi-red commented on main.go:
```diff
@@ -0,0 +1,92 @@
+	}
+
+	// mainで使う環境変数をグローバル変数に代入
+	localPath = os.Getenv("LOCAL_PATH")
+	projectId = os.Getenv("PROJECT_ID")
```
空のときはpanicすると親切そうです

sapphi-red replied:
読み込みは`init`関数という特殊な関数を利用するとよさそうです

sapphi-red commented on pkg/gcp/gcp.go:
```diff
@@ -0,0 +1,113 @@
+	// バケットとメタデータの設定
+	bucket := client.Bucket(bucketName)
+	bucketAtters := &storage.BucketAttrs{
+		StorageClass: "COLDLINE",
+		Location:     "asia-northeast1",
+		// 生成から90日でバケットを削除
+		Lifecycle: storage.Lifecycle{Rules: []storage.LifecycleRule{
+			{
+				Action:    storage.LifecycleAction{Type: "Delete"},
+				Condition: storage.LifecycleCondition{AgeInDays: 90},
+			},
```
日数指定とクラス指定は環境変数に切り出したいです :pray:

logica0419 commented:
それじゃあその方向で行きます

Commit:
レビューの反映と各種微調整
・ツールの名称変更
・モジュールの名前を変更
・バケット名など、各種「traQ」が入っていたところを変更
・環境変数を.envから直接指定に変更
・環境変数をそれぞれのパッケージのinitで読むように
・バケットのクラスと持続日数を環境変数に切り出し
・環境変数がどれか一つでも空ならpanicで終了するように
・上記の変更を受けたREADME.mdの改変
・README.mdにローカルで実行するときのシェルスクリプト例を記載
・log.Fatalだった部分をlog.Printとpanicで表示するように変更

logica0419 commented:
レビューの方反映しました！
再レビューの方よろしくお願いします。
付け加えたものが余計そうならそこも指摘していただけると…

logica0419 committed:
Update README.md

sapphi-red commented on gcp/gcp.go:
```diff
@@ -0,0 +1,123 @@
+package gcp
+
+// GCPのライブラリを使った処理を集めるパッケージ
```
パッケージの説明はこういう風に`package ～`の上に持ってきたほうがよいです
https://golang.org/src/sort/sort.go

sapphi-red commented on gcp/gcp.go:
```diff
@@ -0,0 +1,123 @@
+
+func init() {
+	// 環境変数をグローバル変数に代入
+	localPath = os.Getenv("LOCAL_PATH")
+	gcpKey = os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")
+	projectId = os.Getenv("PROJECT_ID")
+	storageClass = os.Getenv("STORAGECLASS")
+	duration, _ = strconv.ParseInt(os.Getenv("DURATION"), 0, 64)
+
+	if localPath == "" || gcpKey == "" || projectId == "" || storageClass == "" || duration == 0 {
+		log.Print("Error: Failed to load env-vars")
+		panic("empty env-var(s) exist")
+	}
```
ここでは環境変数を読み込まずに、`main`のほうで読み込んで、渡してくるとよさそうです

```go
type GCPClient struct {
  projectId string
}

func NewGCPClient (projectId string) *GCPClient {
  return &GCPClient {
    projectId: projectId,
  }
}

func (c *GCPClient) CreateBucket() {}
```

logica0419 replied:
これは…パッケージで使う環境変数を構造体として定義しといて、それに付随するメソッドとしてパッケージの関数を設定するってことですかね？メインの方で構造体の型の変数を定義してメソッドとして呼び出す感じ？

sapphi-red replied:
そんな感じです
環境変数の読み込みは一箇所にかためておくと、例えば設定ファイルでも設定変えられるようにしたいといった場合に嬉しいです

logica0419 replied:
とりあえず環境変数の読み込みは一括にまとめるということですね。わかりました。
パッケージへの受け渡しは構造体とメソッドが一番綺麗にまとまりそうなのでそれで実装を試みようと思います。

sapphi-red commented on main.go:
```diff
@@ -0,0 +1,85 @@
+	}
+
+	// バケットへファイルをコピー
+	objectNum, errs := gcp.CopyDirectory(*bucket, files)
```
filesではなくて、`fs.FS`型を受け取るとよさそうです
https://golang.org/pkg/os/#DirFS

logica0419 replied:
fs.FS形受け取っても、Openするためにはファイル一つ一つのNameが必要なのでファイルが開けないのではと思ったのですが、fs.FS形のデータのみでファイルを一つずつ開ける方法あります？

sapphi-red replied:
https://golang.org/pkg/io/fs/#ReadDir
```
var dir fs.FS

files := fs.Readdir(dir, ".")
for file := range files {
   f := dir.Open(file)
}
```
こんな感じで動いたりしませんか？(試してないです)

logica0419 replied:
なるほど、そっちの関数で[]DirEntry型にするということですね
わかりました
fs.FSの方が軽いから？で認識はあってますかね

sapphi-red replied:
`fs.FS`を使うほうが抽象化されててgcpパッケージでそれがローカルのファイルストレージであるという具体性を消せるからですね

logica0419 replied:
なるほど、確かに一つ一つのファイルのデータよりディレクトリ全体のデータなので抽象性は増しそうですね
[]DirEntry型を引数として渡していたのはディレクトリを読み込み[]DirEntryにするための関数でerrorが返り値になるので、ディレクトリ読み込みのerrはpanicに入れて終了させ、ファイルコピーのエラーはerrsにスタックさせて表示するのみとしたかったからですね。そこで関数を分けてたので[]DirEntry型の返り値と引数が発生してました。
結局、CopyDirectory関数にディレクトリ読み込みのerrorであるerrとファイルコピーのエラーのスライスであるerrsを両方返り値として設定することで、ディレクトリ読み込みとコピーを一つの関数にまとめました。これで引数はバケットと環境変数のみになります。

sapphi-red commented on webhook/webhook.go:
```diff
@@ -0,0 +1,88 @@
+)
+
+func init() {
+	// 環境変数をグローバル変数に代入
```
こっちも、ここでは環境変数を読み込まずに、`main`のほうで読み込んで、渡してくるとよさそうです

Commit:
レビューを一部反映
・パッケージの説明の場所を移動
・引数に抽象性が無かったため、元々mainパッケージにあったLoadDirectory関数の中身をCopyDirectory関数に移植しまとめた

Commit:
レビューの反映と一部変更
・関数への環境変数の受け渡しを、環境変数を管理する構造体に付属するメソッドとすることで整理
・環境変数でバケットの名前を決められるように
・ログにバケットの名前が流れるように
・バケット名・Webhookメッセージの生成はmainパッケージ担当に

logica0419 commented:
レビューいただいたところまでとりあえずやったので、再レビューをお願いしたいです

sapphi-red commented on gcp/gcp.go:
```diff
@@ -0,0 +1,108 @@
+	"google.golang.org/api/option"
+)
+
+// 環境変数を管理する構造体の定義
+type GcpEnv struct {
+	LocalPath    string
+	GcpKey       string
+	ProjectId    string
+	StorageClass string
+	Duration     int64
+}
```
環境変数のまとまりという概念を構造体で表すのがいいと思うのですが、もしそういう概念なら、メソッドが生えるのは正しくないと思います
`gcpenv.CreateClient()`という書き方になりますが、これは文意がつかめません
なので、そうするなら`CreateClient(gcpEnvs)`になりますが、これも変数をたらい回しにすることになるのであまりよくないです

なので、GCPを扱うものという概念を構造体で表すとよいです。そうすれば、メソッドが生えても自然です

logica0419 replied:
なるほど…扱い方の問題にはなりますが、言われてみれば環境変数にメソッドが生えるのは少し不自然ですね…
名前と扱い方を変えておきます

Commit:
環境変数を管理する構造体から、パッケージを管理する構造体に名称・扱いを変更

Commit:
コメントの変更
「パッケージを管理する構造体」から「パッケージを表す構造体」へ

logica0419 commented:
名称を ( {パッケージ名}Struct ) としました。コメントは少し悩みましたが「パッケージを表す構造体」がなんとなく自分の中ではしっくりきたのでそうしました。少し実態とズレがあるかも？という感じですが。

sapphi-red commented on gcp/gcp.go:
```diff
@@ -0,0 +1,108 @@
+)
+
+// パッケージを表す構造体の定義
+type GcpStruct struct {
+	LocalPath    string
+	GcpKey       string
+	ProjectId    string
+	StorageClass string
+	Duration     int64
+}
```
パッケージを表す構造体っていう見方をするなら、そもそも構造体を使わずにこうするほうが自然じゃないですか？

```go
var (
  localPath string
)

func SetVars(_localPath string) {
  localPath = _localPath
}
```

今考えたいのはそうではなくてGCPを扱う構造体だと思います

---

構造体の型名に`struct`はつけるべきではないです
これはシステムハンガリアン記法が推奨されないことやgoでは短い名前のが好まれるためです
https://golang.org/doc/effective_go#names

logica0419 replied:
確かにパッケージ内のグローバル変数で解決できてしまうんですよね…
「GCPを扱う構造体」がいまいち理解できてない説ありますね…
今思い至りましたが「関数の集まり」のパッケージではなく、「変数とメソッドの集まり」としてのクラス的な考え方でStructを定義したらいい、という事を言ってるので合ってますかね？
名称付けが難しい…どうしたらいいんでしょう、翠さんだったらなんて付けますか？全く別の名前にしたら意味が分からなくなりそうだし…

logica0419 replied:
今システムハンガリアンについて少し調べてきました。
アプリケーションハンガリアンにした方が良いという事ですかね、恐らくは

logica0419 replied:
あとはパッケージで被りが回避できるので、structの型から「gcp」や「webhook」は消し去ってもいいんですかね？

Commit:
構造体の名称と概念を整理
・mainパッケージの「MainEnv」は環境変数を扱うためのグローバル変数のためのメソッドなし構造体
・gcp、webhookパッケージの「Handler」はそれぞれGCP、Webhookのハンドラとしてのメソッド付き構造体
とし、種別を分けた。コメントにもそう記しておいた。

logica0419 commented:
メソッド付きでGCPやWebhookを扱うための構造体は「Handler」とし、パッケージごとに同名で用意するようにしました。
また、そうするとmainパッケージのstructは意味合いが違うなと考えたので、環境変数をグローバル変数として扱いやすい用に設定した変数のまとまりとしての構造体に意味合いを変えておきました。
むしろわかりにくくなるようなら申し訳ないです。とりあえず自分なりに最大限正確な意味づけを行いました。

FujishigeTemma commented on gcp/gcp.go:
```diff
@@ -0,0 +1,108 @@
+	"google.golang.org/api/option"
+)
+
+// GCPを扱う構造体の定義
+type Handler struct {
+	LocalPath    string
+	GcpKey       string
+	ProjectId    string
+	StorageClass string
+	Duration     int64
+}
+
+func (env *Handler) CreateClient() (*storage.Client, error) {
+	// jsonで渡された鍵のサービスアカウントに紐づけられたクライアントを建てる
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx, option.WithCredentialsFile(env.GcpKey))
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("Successfully set a client in \"%s\"", env.ProjectId)
+	return client, err
+}
```
GCPを扱う構造体は、いわゆるクライアントだと思えばよいです。
つまり、既存のクライアントをラップして使いやすいクライアントを作るわけです。
```suggestion
// GCPを扱う構造体の定義
// LocalPathはGCPには関係なくてクライアントを使うときに必要な値なのでここでは使わない。
type Client struct {
	ProjectID    string
	StorageClass string
	Duration     int64
	storage.Client
}

func NewClient(key, projectID, storageClass string, duration int64) (*Client, error) {
	// jsonで渡された鍵のサービスアカウントに紐づけられたクライアントを建てる
	ctx := context.Background()
	client, err := storage.NewClient(ctx, option.WithCredentialsFile(key))
	if err != nil {
		return nil, err
	}

	log.Printf("Successfully set a client in \"%s\"", projectID)
	return &Client{ projectID, storageClass, duration, client }, err
}
```
それで、↑にメソッドを生やせばよいです。
と、ここまで説明してなんですが個人的にはこの大きさのプログラムならパッケージを切る必要もなくてファイルを分けて関数に切り出すぐらいでいいと思ってます。
それで環境変数は`main.go`のグローバル変数に持って関数に渡すようにすれば良いんじゃないかな？

sapphi-red replied:
環境変数は`config`パッケージに`Load`みたいな関数用意して構造体返すようにするのがよさそうだとぼくは思います

logica0419 replied:
確かに！GCPではクライアントが該当しますね。気づかなかった。
んでもって結局感がすごい… コードの分け方とか変数の持ち方・定義の仕方・引数の形とかって人それぞれでやり方ありますよね… わかってはいた
う～ん… 環境設定は構造体と読み込み関数を別ファイルで用意してそいつを引数で渡す感じでまとめようと思います。パッケージまとめるかどうかは今回くらいだとどっちでもいい感じなのかな？プロジェクトのサイズが大きければ分割した方が良いんですね。今回まとめるかはちょっと考えます。

今回とりあえず、最低限
・環境変数は一個の関数でまとめて読み込もう
・メソッドを生やすときは概念的に大きな型にして生やそう
・引数を渡すときは抽象性をできるだけ高めよう
は全体的にマナーとして守ったほうがいいんじゃないか、と思いましたが、これ以外に今まで挙げられた問題点の中で守った方が良い"一般的"な(=人によらない)マナーってありますかね？(Goの基本的な文法以外の理解しやすいコードを書くという面で)

sapphi-red replied:
環境変数を構造体として取りまわすならパッケージ切ったほうがよくて、別にそうしないならmainで読み込んじゃっていいと思ってます
今のコードは構造体として取りまわしてるのでパッケージ切ったほうがいいって言いましたが、大きさ的には別に切り分ける必要はないと思ってます

こういう(=設計とかアーキテクチャって言われる話)のは今じゃなくて将来に焦点を当てて考えるもので、個人間での好みというよりはそのコードの未来に何を見てるかって感じがします

logica0419 replied:
「構造体で取り回すなら切った方がいい」の理由がピンときてないのですがどうしてでしょう？配列的な構造体の使い方だったらいいとかそういうのってあります？
アーキテクチャ、聞いたことはあります！覚えておかないと… まだ「目の前の実装を綺麗にまとめる」までしか考えられてないので未来でどう使われるか考えられるようにならないとなと思います

sapphi-red replied:
> 「構造体で取り回すなら切った方がいい」の理由

これは単純に構造体で取り回すようにするとその関連のコードが増えるからですね

logica0419 replied:
なるほど…？これは多分もっと複雑なコードにならないとありがたみがわからないやつかもしれない…
とりあえず今回はmainパッケージに素でグローバル変数として環境変数置きました

Commit:
環境変数とパッケージの変更
・パッケージをmainに統一
・環境変数はmainパッケージのグローバル変数として読み込み
・関数を少し移動
・GCP strageパッケージの更新

logica0419 commented:
レビュー反映完了しました！確認お願いします

Commit:
README.mdで誤って書き換わっていたところを修正

Commit:
コメントのミスを修正

sapphi-red commented on gcp.go:
```diff
@@ -0,0 +1,98 @@
+	defer writer.Close()
+
+	// 書きこみ
+	_, err = io.Copy(snappyWriter, original)
```
メモリが豊富じゃない環境で動かすので`io.Copy`は避けたいです
https://medium.com/eureka-engineering/file-uploads-in-go-with-io-pipe-75519dfa647b

logica0419 replied:
io.Copyを調べてた時、メモリ食うって書いてたんでやべーかなとは思ってましたがやはり…
Pipeですね、使ってみます

sapphi-red commented on main.go:
```diff
@@ -0,0 +1,88 @@
+	webhookSecret = os.Getenv("TRAQ_WEBHOOK_SECRET")
+
+	// 環境変数がどれか一つでも空だったらエラーを吐いて終了
+	if localPath == "" || gcpKey == "" || projectId == "" || bucketName == "" || storageClass == "" || duration == 0 || webhookId == "" || webhookSecret == "" {
```
空かの判定は関数に切り出すとともに、どの環境変数が空かどうかをエラーに出力するとうれしそうです

logica0419 replied:
そうですね、これは別口にしましょう

sapphi-red reviewed (commented):
どういう構造にするとうれしいかみたいな話は今回だとconoha-backup-helperとこのソースを一つのコードに落とし込みには、みたいなこと考えるといいです(今回は混ぜないほうがいいと思うので混ぜる必要はないです)

Commit:
レビューの反映
・環境変数関連の処理をenv_var.goにまとめた
・関数の読み込みと関数が空かチェックするのを別の関数に
・空の環境変数をエラーログに表示するようにした
・io.Pipeとgoroutineを使った並列処理を一応実装

logica0419 commented:
環境変数関係は一応僕なりに綺麗にまとめました。ダメなところもまだあると思うのでレビューよろしくお願いします。
io.Pipeの方、色々調べながら実装してみて気づきましたが、恐らく今回のケースだとPipe使用してもほどんどメモリ使用量は変わらないんじゃないかと思います。どうやら非同期処理にしてメモリ負荷が軽くなるのは、開いたデータを送るときに変換したくて、同期処理の場合にその変換データをバッファに残さなければいけないときのようですね。今回はfs.File(io.Readerインターフェースとして使用)で開いたファイルを別型に変換しない(snappyはwriterからwriterへの変換)ですし、GCPのWriterはCopyすれば勝手に書き込んでくれるので、メモリを食わないということになります。goroutineとio.Pipeの勉強になったのでそこはすごく良かったですが、今回機能としては多分いらないです。
実際にprofileパッケージ(runtime/pprofを使いやすくした版)を使ってメモリプロファイル取った結果がこれです。上がio.Pipe + goroutine使用時、下がio.Copyのみの時になります。今回最大で150MBの動画ファイルを送信しましたが、どちらも16MBちょっとまでしかメモリは使われませんでした。正直このプログラムの場合、Pipeを使っても内部的には全く同じ処理になるんですね…。Readerで取れたのをそのまま送れるので、下手にこねくり回さない方が処理も軽くなります。
もし違う部分に使うことを言っているのであれば大変申し訳ないです。一応conoha-backup-helperの実装とほぼ同じことをやっているので翠さんが言ってるのは今回の改変みたいなことだと思うんですよね…。conoha-backup-helperの時の(*Connection) ObjectGetの場合、中身が渡されたio.Writer型の変数に書き込まれるのでio.Readerに変換する際io.Pipeが非常に有用でした。ですが、今回ローカルファイルがもともとio.Readerで読み込めているため、これ以上効率化は成しえないと僕は思っています。
io.Pipeとio.Copyに関して図形式のわかりやすい資料もあったので載っけときます
https://christina04.hatenablog.com/entry/golang-io-package-diagrams
こちらは初心者なので、説明に抜けがあると納得していただけないと思った結果、長文になってしまい大変失礼しました。納得していただけたらio.Pipe使った処理は消したいのですが、どうでしょうか？
![](https://user-images.githubusercontent.com/84759675/125852586-06d98e01-f5c6-42cd-a25a-b38bc4d1e2e2.png)
![](https://user-images.githubusercontent.com/84759675/125852590-fe9b8785-2871-4175-9350-4c6dc93a2fa2.png)

mazrean commented:
たぶんlogicaくんの言っているのであっている(`io.Pipe`が活躍するのは基本`io.Reader,Writer`を関数の引ううで渡して中で何か書き込まれる時なので)

logica0419 commented:
> どういう構造にするとうれしいかみたいな話は今回だとconoha-backup-helperとこのソースを一つのコードに落とし込みには、みたいなこと考えるといいです(今回は混ぜないほうがいいと思うので混ぜる必要はないです)

そうだ、これに関して今考えが至りました
アーキテクチャを考える時、「データ保持やパッケージ分割の規格を他のプロジェクトと共通化して、可読性や汎用性を上げる」という理解をしましたがこの理解でどうでしょうか？

sapphi-red commented:
`io.Pipe`の話はそうです
ぼくの勘違いです
ごめんなさい

sapphi-red commented:
> 「データ保持やパッケージ分割の規格を他のプロジェクトと共通化して、可読性や汎用性を上げる」

共通化するというよりは何かがあったときに共通化しやすい形にする、のが近いですね
今じゃなくて将来のことを見てるので

logica0419 commented:
> > 「データ保持やパッケージ分割の規格を他のプロジェクトと共通化して、可読性や汎用性を上げる」
> 
> 
> 
> 共通化するというよりは何かがあったときに共通化しやすい形にする、のが近いですね
> 
> 今じゃなくて将来のことを見てるので
> 
> 

なるほど:kanzenrikai:しました
将来を見据えるというのはこのような例があるんですねぇ…勉強になります

Commit:
io.Pipeを使用しないコードに戻した・環境変数読み込んだらログを表示するように

logica0419 commented:
というわけでio.Pipeの修正とちょっとログ追加しました
恐らく最終版…？と思っています、レビューをお願いします
そろそろOKをもらえるようなコードになってれば嬉しいです…

Commit:
バグの修正
writerをsnappyWriterより先にCloseしていたことによる、0KBファイルが生成される問題を修正
(deferはLast In, First Outでスタックした処理を実行することを失念していた)

sapphi-red commented on env_var.go:
```diff
@@ -0,0 +1,61 @@
+
+func EnvVarLoad() {
+	// 環境変数を取得
+	localPath = os.Getenv("LOCAL_PATH")
```
```go
func getEnv(name string) string {
  value := os.Getenv(name)
  if value == "" {
    log.Println(name)
    panic()
  }
  return value
}
```
みたいなの定義して使うのでも十分そうな気がします

全部のエラーをちゃんと出すなら最初からライブラリ使う方がいいと思うので
https://github.com/spf13/viper
これとか

logica0419 replied:
あ、こんなアプリ設定をやりやすくするモジュールあるんですね…世の中何でもあるもんだ
そうですね、関数組んで個別でエラー出す形にしちゃいますね
GitHub Actions未履修なので参考資料見ながらやってみます

sapphi-red reviewed (commented):
ビルド結果をリリースにくっつけるgithub actionsをこれを参考に追加してほしいです
https://github.com/traPtitech/conoha-backup-helper/blob/master/.github/workflows/release.yml

logica0419 committed:
レビューの反映
・環境変数の一つあたりの読み込みとチェックを一つの関数としてくくった
・モジュールの更新

FujishigeTemma commented:
今、グローバル変数を直接使ってると思うんですけど、引数で取るようにしてほしいです。
```
var name string
func main() {
  name = getEnv()
  upload()
}
func upload() {
  client.backup(name)
}
```
みたいなコードを
```
func main() {
  name := getEnv()
  upload(name)
}
func upload(name string) {
  client.backup(name)
}
```
みたいに書き換えるってことです。

logica0419 commented:
実は最初環境変数は引数として扱っていて、その後このmainパッケージ全体で使うのでパッケージ内のグローバルとして定義するのが良いという判断をしたのですが、グローバル変数だと何がマズいと考えていますか？引数とすることのメリットまたはグローバル変数の懸念点を教えていただきたいです。

FujishigeTemma commented:
変数の影響の伝播が意味的に追いかけにくいことですね。

logica0419 commented:
う～んなるほどわかりました
constにすればいいという話ではなく、環境変数変えたときにどこが変わるかわかりにくいよね、という話だという認識で合ってますかね？

FujishigeTemma commented:
あってます:ok_hand:

logica0419 commented:
~うへぇ～変数全部書くのめんD～~
頑張ります

logica0419 committed:
環境変数を、すべて引数として受け渡すように

logica0419 commented:
環境変数の扱いは変更しました
あと今更思ったんですがsnappyで圧縮してるのに末尾に.snapって付けなくていいんですか？実際バックアップの復元が必要になったときにどう扱うのかが気になってて

sapphi-red commented:
どう扱うかってのがよくわかんないですが、普通に展開すればいいだけだと思います
```
$ go install github.com/golang/snappy/cmd/snappytool@v0.0.4
$ cat a | snappytool -d > b
```
こんな感じに

logica0419 commented:
あ、拡張子そのままで展開できるツールあるんですね…なるほど
扱うってのはただ展開して戻すってことが言いたかっただけですね
Windows上のsnappy解凍ソフトでは扱えなくてきちんと解凍できるのかわからなかったので、ちょっとUbuntu上で試してきます

logica0419 commented:
@sapphi-red 
catでファイル書き込むときに`snappytool：command not found`って言われて空のファイルが出力されてしまうんですが、これどこのディレクトリでやらなきゃいけないとかあります？それともインストールの仕方が間違ってる？

sapphi-red commented:
GOBINにパスが通ってないんだと思います

logica0419 commented:
う～んGOBINにパス通してそこにsnappytoolのバイナリファイルあるのも確認しましたがダメですね…
![image](https://user-images.githubusercontent.com/84759675/126893803-ca22a77b-f237-47a4-a215-8d0c8dd8f38f.png)

sapphi-red commented:
環境変数PATHにGOBINは含まれてますか？(スクショに写ってないので一応)

logica0419 commented:
含まれてると思います…
![image](https://user-images.githubusercontent.com/84759675/126907456-7b5109b8-2d35-4fa7-9a17-6e7364b51607.png)

sapphi-red commented:
それはGOBINの設定であってPATHではなくないですか？

logica0419 commented:
あ～なるほど、Goの環境変数を通すだと思ってました…
今調べて環境変数、PATHについて理解してきました…。Linuxの理解が足りておらず、大変申し訳ないです。
今実行はできましたが今度はsnappyにcorrupt inputだと言われて頭を抱えております…
調べてみましたが、Block型/Stream型の区分が一致していない説が浮上しているくらいですね… ([これ](https://gowalker.org/github.com/golang/snappy)の最初の"There are actually two Snappy formats: block and stream"の部分)
![image](https://user-images.githubusercontent.com/84759675/126918300-4db94735-e7d1-4567-bb77-1d96ec4218f1.png)

logica0419 commented:
とりあえずReaderを使ったデコーダーを実装したところ無事解凍できました…お騒がせしてすみません。
解凍するときいちいちこれ書くのめんどいな～と思ったんですがsnappytoolにStreamでデータを扱う機能ってあるんですかね？(snappytoolのドキュメント見つからなかった)
--help見た感じなさそうですが、便利ツールとしてdecoder上げといた方がいいですか？超適当実装なので色々とガッバガバですけど (既にやってたら申し訳ない)
```golang decoder.go
package main

import (
	"io"
	"log"
	"os"

	"github.com/golang/snappy"
)

var (
	encodedDir string
	decodedDir string
)

func init() {
	encodedDir = os.Getenv("ENCODED_DIR")
	decodedDir = os.Getenv("DECODED_DIR")
	if encodedDir == "" || decodedDir == "" {
		log.Print("Error: empty env-var(s) exists")
		panic("empty env-var(s) exists")
	}
}

func main() {
	// ローカルのディレクトリ構造を読み込み
	files, err := os.ReadDir(encodedDir)
	if err != nil {
		log.Print("Error: failed to read directory")
		panic(err)
	}

	for _, file := range files {
		// ローカルのファイルを開くためのReaderを用意
		reader, err := os.Open(encodedDir + "/" + file.Name())
		if err != nil {
			log.Printf("Error: failed to open file: %s", file.Name())
			panic(err)
		}
		// それを元にSnappyの解凍用Readerを用意
		snappyreader := snappy.NewReader(reader)

		// 新しいファイルを作成
		decodedFile, err := os.Create(decodedDir + "/" + file.Name())
		if err != nil {
			log.Print("Error: failed to create a new file")
			panic(err)
		}
		defer decodedFile.Close()

		// 書き込み
		_, err = io.Copy(decodedFile, snappyreader)
		if err != nil {
			log.Print("Error: failed to write to a new file")
			panic(err)
		}

		log.Printf("Successfully decompressed %s", file.Name())
	}
}
```

sapphi-red commented:
```shell
$ go install github.com/golang/snappy/cmd/snappytool@35a8406
```
でできそうな感じがします

logica0419 commented:
本当だ、最新バージョンでBlock型を扱うように変更されたみたいですね… コミットログ見てなかったので気づかなかったです… 誠にありがとうございます
snappytoolを使って無事変換できました！大変ありがとうございます。無事に圧縮できていることが分かったのでこれで自信をもって運用していただけます。というわけで問題なければレビューでapprove・マージしていただければ幸いです。

logica0419 committed:
.gitignoreの更新、モジュールの更新

sapphi-red commented on main.go:
```diff
@@ -0,0 +1,60 @@
+
+	// bucketName + バックアップ日 をバケット名とし、bucketNameに再代入
+	t := &startTime
+	bucketName = fmt.Sprintf("%s-%d-%d-%d", bucketName, t.Year(), t.Month(), t.Day())
```
そういえばこれmod nでサイクルするようにしてほしいです
過去n個のバックアップが残るように
「b-1に書き込む、b-2に書き込む、b-3に書き込む、b-1に書き込む、b-2に書き込む、b-3に書き込む…」
みたいな感じです
意図はバケットが増えていかないようにするためです

logica0419 replied:
削除までの期間を環境変数で設定できるようにしていますが、定期的なバックアップならそれを適切に設定することで無限バケットを回避できるのではないですか？GCPのdurationって切れても残ります？
また、同じバケット名で書き込んでいくなら日付は消えることになりますよね？最後に書き込んだバケットがわかりづらくてどれが最新かわからなくなるというデメリットも出るんじゃないかと考えましたがどうですか？

logica0419 replied:
今のバケット一覧を取得してもしn個以上だったら古いものから削除するとかなら多分デメリットは回避できて実装できそうな範囲内なんですが、durationの挙動が僕の意図しているものであればわざわざやる意味が無いかな～というのが僕の意見です…

sapphi-red replied:
オブジェクトは消えるんですが、バケットが消えないんです

logica0419 replied:
マジですかそれ？泣いてます
わかりました、リスト取得して古い方からバケットを抹消するようにプログラム組んでみます

sapphi-red replied:
更新日時は見ればわかるので、そこまでしなくても月のmodとかで十分です

logica0419 replied:
あ〜〜なるほど、承知しました、月1くらいでいいんですね…
ではご指定の通りに実装してみます

logica0419 committed:
レビューの反映
・バケットの名前がmod nでサイクルするように(環境変数でnは指定)
・同じ名前のバケットがある場合のエラー処理を別枠に
・Webhookメッセージやログを少し修正

logica0419 commented:
パッと最低限必要そうなのを実装しました。こんなもんでどうでしょうか？
もしオブジェクトが残っていた場合に削除するとかが必要そうなら言ってください(環境変数の指定さえ間違えなければそんなことにはならないけども)

sapphi-red commented on gcp.go:
```diff
@@ -0,0 +1,103 @@
+func CreateBucket(client storage.Client, projectId string, storageClass string, duration int64, bucketName string) (*storage.BucketHandle, error) {
+	// バケットとメタデータの設定
+	bucket := client.Bucket(bucketName)
+	bucketAtters := &storage.BucketAttrs{
```
```suggestion
	bucketAttrs := &storage.BucketAttrs{
```

logica0419 replied:
うわ本当だタイポしてる、直します

logica0419 replied:
本当だタイポしてる
直しました

logica0419 committed:
attrsのタイポミスの修正

sapphi-red commented on gcp.go:
```diff
@@ -0,0 +1,103 @@
+	if err != nil {
+		// バケットが既にある場合のエラー(409: Conflict)を別枠で処理
+		if strings.Contains(err.Error(), "Error 409") {
+			return bucket, errors.New("Error 409")
```
エラー返さずにここでログだしてよさそうです

sapphi-red commented on main.go:
```diff
@@ -0,0 +1,64 @@
+	defer client.Close()
+
+	// bucketName + 月(mod n) をバケット名とし、bucketNameに再代入
+	t := &startTime
+	bucketName = fmt.Sprintf("%s-%d", bucketName, t.Month()%time.Month(bucketNum))
```
```suggestion
	bucketName = fmt.Sprintf("%s-%d", bucketName, startTime.Month()%time.Month(bucketNum))
```

sapphi-red commented on main.go:
```diff
@@ -0,0 +1,64 @@
+	// バケットへディレクトリをコピー
+	objectNum, err, errs := CopyDirectory(*bucket, localPath)
+	if err != nil {
+		log.Print("Error: failed to load local directory")
```
ここでエラーがかえったからと言って、「ディレクトリのコピー」の処理の中の「ディレクトリのロード」のエラーであることは関数の中身を知らない限りわからないと思うので、

```suggestion
		log.Print("Error: failed to copy directory:", err)
```

のがよいです
こうすれば関数の実装が変わっても不整合が発生しません
(ほかの箇所も`err`を出力したほうがいいです)

logica0419 replied:
それでもいいんですけど、ファイルコピー中にエラーが発生した場合でもpanicすることになるんですが、それでいいんですか？
Conohaのオブジェクトストレージのバックアップの方ではいくつかエラーが発生しても残りのファイルはバックアップしておこう的な発想だと思ったんですが
やるとしたらerrsにappendして返すってのは考えはしましたが

sapphi-red replied:
> ファイルコピー中にエラーが発生した場合でもpanicすることになる

エラーログの内容を変えることはロジックには影響しないと思います
(`err`と`errs`をまとめるべきっていう意図ではないです)

logica0419 replied:
あ～なるほど
この後他の要因でpanicしたいと思った時でも大丈夫なようにってことですね、ようやく理解しました
理解力無くてすみません

logica0419 replied:
直ってなかった
>(ほかの箇所もerrを出力したほうがいいです)

に関して
panicでエラー出してますけどprintでも出したほうがいいですかね？その方が見やすい？

logica0419 replied:
panicの中身って別の文字列の方がいいんでしょうか？もしくは全部panicの中にまとめちゃった方がいい？

sapphi-red replied:
あんまり考えてなかったんですが、`panic`側にまとめたほうがいいですね
そうすれば`stderr`に流れるので(`print`だと`stdout`にいくはず)

logica0419 replied:
fmt.Errorf使ってpanic側にまとめることにします

logica0419 committed:
レビューの反映

logica0419 commented:
パパっと直しました、よろしくお願いします

logica0419 committed:
エラーメッセージをpanicに集約

logica0419 commented:
修正終わりました～よろしくお願いします

sapphi-red commented on main.go:
```diff
@@ -0,0 +1,54 @@
+	// バケットへディレクトリをコピー
+	objectNum, err, errs := CopyDirectory(*bucket, localPath)
+	if err != nil {
+		panic(fmt.Errorf("Error: failed to copy directory - %s", err))
```
動作的には同じなのでどっちでもよいといえばよいんですがここで`fmt.Errorf`を使うのは不自然です
というのも`fmt.Errorf`は`error`型を返すのであって、文字列型を返しません
ただこの場合は`panic`が`interface{}`型を受け取るので、結果的に`error`型が文字列型に変換されて同じ動作になっています

logica0419 replied:
panicはerror型を受け取るもんだと勝手に思い込んでたんですがstring型だったんですね…
わかりました、stringにしますね

logica0419 committed:
panicにstring型を渡すよう修正

sapphi-red reviewed (approved):
LGTM 🎉

logica0419 commented:
とりあえずLGTMが出て、その後追加の修正が来ないのでマージしますね～
